File: LeetCodePractice/src/bst/AllPossibleFullBinaryTrees.java
package bst;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/*A full binary tree is a binary tree where each node has exactly 0 or 2 children.

Return a list of all possible full binary trees with N nodes.  Each element of the answer is the root node of one possible tree.

Each node of each tree in the answer must have node.val = 0.

You may return the final list of trees in any order.

Example 1:

Input: 7
Output: [[0,0,0,null,null,0,0,null,null,0,0],[0,0,0,null,null,0,0,0,0],[0,0,0,0,0,0,0],[0,0,0,0,0,null,null,null,null,0,0],[0,0,0,0,0,null,null,0,0]]
Explanation:

Note:

1 <= N <= 20*/
public class AllPossibleFullBinaryTrees {
	Map<Integer, List<TreeNode>> cache = new HashMap<>();

	public List<TreeNode> allPossibleFBT(int N) {
		List<TreeNode> res = new ArrayList<>();
		if (N % 2 == 0) {
			return res;
		}
		if (N == 1) {
			TreeNode root = new TreeNode(0);
			res.add(root);
			return res;
		}
		if (cache.containsKey(N)) {
			return cache.get(N);
		}
		N--;
		for (int i = 1; i < N; i += 2) {
			List<TreeNode> leftPool = allPossibleFBT(i);
			List<TreeNode> rightPool = allPossibleFBT(N - i);
			for (TreeNode left : leftPool) {
				for (TreeNode right : rightPool) {
					TreeNode root = new TreeNode(0);
					root.left = left;
					root.right = right;
					res.add(root);
				}
			}
		}
		cache.put(N, res);
		return res;
	}
}

File: LeetCodePython/AllPossibleFullBinaryTrees.py
"""
Given an integer n, return a list of all possible full binary trees with n nodes. Each node of each tree in the answer must have Node.val == 0.

Each element of the answer is the root node of one possible tree. You may return the final list of trees in any order.

A full binary tree is a binary tree where each node has exactly 0 or 2 children.

Example 1:
Input: n = 7
Output: [[0,0,0,null,null,0,0,null,null,0,0],[0,0,0,null,null,0,0,0,0],[0,0,0,0,0,0,0],[0,0,0,0,0,null,null,null,null,0,0],[0,0,0,0,0,null,null,0,0]]

Example 2:
Input: n = 3
Output: [[0,0,0]]


Constraints:
1 <= n <= 20
"""
from typing import Optional, List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    cache = {}

    def allPossibleFBT(self, n: int) -> List[Optional[TreeNode]]:
        res = []
        if n % 2 == 0:
            return res
        if n == 1:
            res.append(TreeNode(0))
            return res
        if n in Solution.cache:
            return Solution.cache[n]
        n -= 1
        for i in range(1, n, 2):
            left_pool = self.allPossibleFBT(i)
            right_pool = self.allPossibleFBT(n - i)
            for left in left_pool:
                for right in right_pool:
                    root = TreeNode(0)
                    root.left = left
                    root.right = right
                    res.append(root)
        Solution.cache[n] = res
        return res
